test/test_mahalanobis_mixin.py

```diff
@@ -654,6 +693,31 @@ def test_singular_array_init_or_prior(estimator, build_dataset, w0):
       assert str(raised_err.value) == msg
 
 
+@pytest.mark.parametrize('w0', [1e-20, 0., -1e-20])
+def test_singular_array_init_of_non_strict_pd(w0):
+    """Tests that when using a custom array init, it returns the
+    appropriate warning if it is singular. Also checks if the returned
+    inverse matrix is finite. This isn't checked for model fitting as no
+    model curently uses this setting.
+    """
+    rng = np.random.RandomState(42)
+    X, y = shuffle(*make_blobs(random_state=rng),
+                   random_state=rng)
+    P = ortho_group.rvs(X.shape[1], random_state=rng)
+    w = np.abs(rng.randn(X.shape[1]))
+    w[0] = w0
+    M = P.dot(np.diag(w)).dot(P.T)
```

Comment:
this seems a bit convoluted just to generated a singular matrix? You could just draw a rectangular matrix `L` and construct `M` as `L.T.dot(L)`?

Reply:
I agree with you, just did it that way because it was done like that on `test_singular_array_init_or_prior_strictpd`. But also this way you have only one zero eigenvalue and control its value directly with the parametrization.

Reply:
Sounds good